finetuning-llama-7b-on-l4/README.md

```diff
@@ -0,0 +1,381 @@
+      containers:
+      - name: loader
+        image: python:3.11
+        command:
+        - /bin/bash
+        - -c
+        - |
+          cat > download-model.py << EOF
+          from huggingface_hub import snapshot_download
+          model_id="meta-llama/Llama-2-7b-hf"
+          snapshot_download(repo_id=model_id, local_dir="/gcs-mount/llama2-7b",
+                            local_dir_use_symlinks=False, revision="main",
+                            ignore_patterns=["*.safetensors", "model.safetensors.index.json"])
+          EOF
+          pip install huggingface_hub
+          python3 download-model.py
+        imagePullPolicy: IfNotPresent
+        env:
+        # kubectl create secret generic l4-demo --from-literal="HF_TOKEN=$HF_TOKEN"
+        - name: HUGGING_FACE_HUB_TOKEN
+          valueFrom:
+            secretKeyRef:
+              name: l4-demo
+              key: HF_TOKEN
+        volumeMounts:
+        - name: gcs-fuse-csi-ephemeral
+          mountPath: /gcs-mount
+      serviceAccountName: l4-demo
+      volumes:
+      - name: gcs-fuse-csi-ephemeral
+        csi:
+          driver: gcsfuse.csi.storage.gke.io
+          volumeAttributes:
+            bucketName: $BUCKET_NAME
+            mountOptions: "implicit-dirs"
+EOF
+```
+
+Give it a minute to start running, once up you can watch the logs of the job by running:
+```bash
+kubectl logs -f -l job-name=model-loader
+```
+
+Once the job has finished you can verify the model has been downloaded by running:
+```bash
+gcloud storage ls -l gs://$BUCKET_NAME/llama2-7b/
+```
+
+Let’s write our finetuning job code by using the HuggingFace library for training.
+
+Create a file named `finetune.py` with the following content:
+```python
+from pathlib import Path
+
+from datasets import load_dataset, concatenate_datasets
+import transformers
+from transformers import AutoTokenizer, AutoModelForCausalLM, TrainingArguments
+from peft import get_peft_model, LoraConfig, prepare_model_for_kbit_training
+import torch
+
+# /gcs-mount will mount the GCS bucket created earlier
+model_path = "/gcs-mount/llama2-7b"
+finetuned_model_path = "/gcs-mount/llama2-7b-american-stories"
+
+tokenizer = AutoTokenizer.from_pretrained(model_path, local_files_only=True)
+model = AutoModelForCausalLM.from_pretrained(
+            model_path, torch_dtype=torch.float16, device_map="auto", trust_remote_code=True)
+
+dataset = load_dataset("dell-research-harvard/AmericanStories",
+    "subset_years",
+    year_list=["1809", "1810", "1811", "1812", "1813", "1814", "1815"]
+)
+dataset = concatenate_datasets(dataset.values())
+
+if tokenizer.pad_token is None:
+    tokenizer.add_special_tokens({'pad_token': '[PAD]'})
+    model.resize_token_embeddings(len(tokenizer))
+
+data = dataset.map(lambda x: tokenizer(
+    x["article"], padding='max_length', truncation=True))
+
+lora_config2 = LoraConfig(
+ r=16,
+ lora_alpha=32,
+ lora_dropout=0.05,
+ bias="none",
+ task_type="CAUSAL_LM"
+)
```

Comment:
Are `target_modules` needed, or is there a sensible default? This article https://www.anyscale.com/blog/fine-tuning-llms-lora-or-full-parameter-an-in-depth-analysis-with-llama-2 has some recommend hyperparametrs to consider optimizing this.

Reply:
I think it's better to not set so people can take this job and use it for other models as well. The default looks works for various models but once you set target_modules to something special you need to adjust it for different models. Cool article you shared though, I wasn't aware that some people seem to get better results by targeting more modules.

I think sticking with the sensible default is best for now. For example, this same code could be used to finetune falcon model, but once you hard code the target_modules it won't work for falcon anymore.